src/traces/bar/plot.js

```diff
@@ -178,12 +201,19 @@ function plot(gd, plotinfo, cdModule, traceLayer, opts) {
                 y1 = fixpx(y1, y0);
             }
 
-            Lib.ensureSingle(bar, 'path')
+            var sel = transition(Lib.ensureSingle(bar, 'path'), opts, makeOnCompleteCallback);
+            sel
                 .style('vector-effect', 'non-scaling-stroke')
-                .attr('d', isBlank ? 'M0,0Z' : 'M' + x0 + ',' + y0 + 'V' + y1 + 'H' + x1 + 'V' + y0 + 'Z')
+                .attr('display', function() { return isBlank ? 'none' : null;})
```

Comment:
Interesting move.

Do you have an example of a bar transitioning in and out of `isBlank`?

Reply:
I should mention here that this was necessary otherwise bars transitioning from `isBlank` to a given value would start growing from the top left corner of the figure (starting at 0, 0)! We need to have blank bars at the right place for transitions in and out of `isBlank` to work. However, even though we have them in the SVG, they shouldn't appear (or we break several mocks) which is why we set them to `display: none`.

> Do you have an example of a bar transitioning in and out of isBlank?

Let me produce a Codepen for you

Reply:
Here's the requested Codepen: https://codepen.io/antoinerg/pen/PoYRBzw

As you can see, there is a problem going from a finite value to blank. The bar disappears right away! Will need to fix this. Very nice catch :male_detective: @etpinard

Reply:
Fixed in https://github.com/plotly/plotly.js/pull/4180/commits/495d036d44636923d9e6d55d7261629c8c7a81ac

Reply:
Oh I see what your doing. Setting a blank path `M0,0Z` breaks the transition, but setting a `y0 = y1` (or `x0 = x1`) rectangle would end up showing the `marker.line` around the bar (which we don't want to do for blank bars)

That `display` trick is clever, but unfortunately it doesn't work under the `frame: {redraw: false}` option: https://codepen.io/etpinard/pen/bGbvZxq - see the `marker.line` is there after the transition is completed.

Reply:
For reference, using the current 

```js
sel.attr('d', isBlank ? 'M0,0Z' : 'M' + x0 + ',' + y0 + 'V' + y1 + 'H' + x1 + 'V' + y0 + 'Z')
```

we get:

![Peek 2019-09-10 15-51](https://user-images.githubusercontent.com/6675409/64645440-e9cb5f80-d3e2-11e9-8f13-723d35b310fe.gif)

Reply:
> That `display` trick is clever, but unfortunately it doesn't work under the `frame: {redraw: false}` option: https://codepen.io/etpinard/pen/bGbvZxq - see the `marker.line` is there after the transition is completed.

I was unaware of that `redraw` thingy. It didn't actually make sense to me why https://github.com/plotly/plotly.js/commit/495d036d44636923d9e6d55d7261629c8c7a81ac would work :laughing: 

I think the solution will be to set `display: none` on transition end and interupt.

Reply:
The solution might be to make

https://github.com/plotly/plotly.js/blob/9b5f0898e13c24f2ad23c9745cac07e48c0e35d9/src/components/drawing/index.js#L445

```js
sel.style('stroke-width', (d.isBlank ? 0 : lineWidth) + 'px');
```

while keeping your new:

https://github.com/plotly/plotly.js/blob/cc871cdb5292bf7c3e75431d3936c2229966cb37/src/traces/bar/plot.js#L141-L142

and dropping the `attr('display', /**/)` lines

Reply:
> The solution might be to make
> 
> https://github.com/plotly/plotly.js/blob/9b5f0898e13c24f2ad23c9745cac07e48c0e35d9/src/components/drawing/index.js#L445
> 
> ```js
> sel.style('stroke-width', (d.isBlank ? 0 : lineWidth) + 'px');
> ```
> 
> while keeping your new:
> 
> https://github.com/plotly/plotly.js/blob/cc871cdb5292bf7c3e75431d3936c2229966cb37/src/traces/bar/plot.js#L141-L142
> 
> and dropping the `attr('display', /**/)` lines

That seemed like a really good idea so I went with it. Unfortunately, the `stroke-width` also transitions which probably isn't what we want. Can you think of another way than setting `display:none` on `transition.end` or `transition.interrupt`?

Thank you @etpinard !

Reply:
> Unfortunately, the `stroke-width` also transitions which probably isn't what we want.

I don't see a problem with that. I mean the `stroke-width` of blank pt is effectively 0, right?

Maybe @archmoj has something to say on this topic.

> Can you think of another way than setting `display:none` on `transition.end` or `transition.interrupt`?

This sounds very hacky and bug-prone to me. I would prefer fixing this problem at the tween level.

Reply:
> I don't see a problem with that. I mean the `stroke-width` of blank pt is effectively 0, right

Not necessarily, you can have a large `stroke-width` with a `null` value and that qualifies as a blank.

Here's a [Codepen showing the fix you proposed](https://codepen.io/antoinerg/pen/zYOWQJz) built off https://github.com/plotly/plotly.js/commit/bda91f7b373e500434c2d884989da33071a8491d. As you can see, the bar has a decreasing line width as it goes to `null` even though we did not explicitly ask to transition that property. One may argue this is good enough but it's not perfect. Let me know what you think!

Reply:
> Not necessarily, you can have a large `stroke-width` with a `null` value and that qualifies as a blank.

Sure, but the `stroke-width` does not render for blank pts e.g 

```js
  Plotly.newPlot(gd, [{
    type: 'bar',
    x: ['A', 'B', 'C'],
    y:  [null, 5, 3];
    marker: {
      line: {
        width: 10
      }
    }
  }])
```

shows up as:

![image](https://user-images.githubusercontent.com/6675409/64652670-6b76b980-d3f2-11e9-89f1-5e84a019296a.png)

i.e. the left-most bar has (effectively) no `marker.line.width`

Reply:
If someone complains about this, we could potentially add a custom `.attrTween` function for `stroke-width`, but I honestly don't see the need to spend time on this at this time.

Reply:
Sounds good! I cherry-picked that fix since you approve of it :)

Reply:
Nicely done! Resolved.